futures-02.md

```diff
@@ -0,0 +1,396 @@
+- Feature Name: futures-2
+- Start Date: 2018-01-22
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Release a breaking change to the `futures` crate.
+The main goals of this breakage are:
+- Learnability and readability improvements acheived by making task wakeup
+  handles into explicit arguments.
+- Separation of `futures` into multiple, smaller crates, which are reexported
+  in a `futures` facade crate. This will make it possible to gradually evolve
+  `futures` libraries without major ecosystem breakage.
+- Removal of deprecated functionality.
+- Minor API cleanup (renamings, generalizations, etc.).
+
+This breakage may (depending on time to accept and implement) overlap with
+library changes resulting from the release of the `tokio` crate
+(formerly `tokio-core`) and `mio` 0.7.
+The `tokio` changes will result in changes to many popular libraries
+in the Rust async ecosystem. By releasing `futures` 0.2 at or around the same
+time, we hope to minimize the number of breaking changes to the ecosystem.
+
+# Motivation
+[motivation]: #motivation
+
+## Explicit Task Wakeup Handles
+
+The `futures` task wakeup mechanism is [a regular source of confusion for new
+users][explicit task issue]. Looking at the signature of `Future::poll`, it's
+hard to tell how `Future`s work:
+
+```rust
+fn poll(&mut self) -> Poll<Self::Item, Self::Error>;
+```
+
+`poll` takes no arguments, and returns either `Ready`, `NotReady`, or an error.
+From the signature, it looks as though the only way to complete a `Future` is
+to continually call `poll` in a loop until it returns `Ready`. This is
+incorrect.
+
+Furthermore, when implementing `Future`, it's necessary to schedule a wakeup
+when you return `NotReady`, or else you might never be `poll`ed again. The
+trick to the API is that there is an implicit `Task` argument to `Future::poll`
+which is passed using thread-local storage. This `Task` argument must be
+`notify`d by the `Future` in order for the task to awaken and `poll` again.
+
+Once users become aware of the TLS `Task`, it's rare that this implicit argument
+is the source of bugs. Storing the `Task` in TLS is also slightly more
+ergonomic when writing manual `Future` implementations since you don't have to
+manually plumb through a wakeup handle.
+
+However, as described earlier, this ergonomic benefit comes at a significant
+learnability cost. Additionally, the implicit argument makes it more difficult
+to tell which functions aside from `poll` will use the `Task` handle to
+schedule a wakeup. Furthermore, the wakeup handle never shows up in async/await
+or combinator code.
+
+This RFC proposes to resolve these issues by making the `Task` wakeup handle
```

Comment:
We've had this debate plenty of times, and I suspect it's already decided, but just because:

I think the "learnability cost" is more of an argument coming from people who have not spent much time writing `impl Future for T`. Those who haven't, when they look at an implementation of `Future`, and missing context, don't like that it's not obvious how the future will be woken up.

Once someone actually does ingest how the TLS task system works, many have noted it's rather elegant, instead.

Crucially, there is a difference between `Future`s implementors and users. Users don't `impl Future`, but instead will be able to just use combinators and `await`. Implementors must write a lot of `impl Future`/`impl Stream`/etc, and having to pass around an additional argument to child futures is just an annoyance, rather than a help.

Reply:
I'm not sure I would say "users don't `impl Future`", I generally recommend to implement your own futures for semantic actions instead of handing around very non-descriptive types with many combinators inside.

Reply:
Context in TLS helps implementors. If one has to weigh towards users or implementors, we should be picking the users.

Also TLS is more expensive than passing an arg.

Reply:
The `#![no_std]` argument is compelling for me- even though TLS is less work once you understand the system, and even assuming the performance is fine (I expect it is), it would be useful for futures to work in that context.

Reply:
>  If one has to weigh towards users or implementors, we should be picking the users.

Somewhat, yes. Weight *more* towards ease-of-use for users. However, if things are too cumbersome for implementors, you'll have libraries that just skip providing `Future` support. As mentioned elsewhere in this thread, having to update `h2` (or `hyper`) will be major sadfaces.

Compared to the possibility that documentation could clean up the confusion for less-informed users who just want to implement a few tiny `Future`s, I'm not sure the weights balance.